hack/generator/pkg/codegen/pipeline_add_cross_resource_references.go

```diff
@@ -0,0 +1,181 @@
+		"addCrossResourceReferences",
+		"Replaces cross resource references with genruntime.ResourceReference",
+		func(ctx context.Context, definitions astmodel.Types) (astmodel.Types, error) {
+
+			result := make(astmodel.Types)
+			knownReferences := newKnownReferencesMap(configuration)
+			isCrossResourceReference := func(typeName astmodel.TypeName, prop *astmodel.PropertyDefinition) bool {
+				ref := referencePair{
+					typeName: typeName,
+					propName: prop.PropertyName(),
+				}
+				_, isReference := knownReferences[ref]
+
+				if doesPropertyLookLikeARMReference(prop) && !isReference {
+					klog.V(0).Infof("\"%s.%s\" looks like a resource reference but was not labelled as one", typeName, prop.PropertyName())
+				}
+
+				return isReference
+			}
+
+			visitor := makeCrossResourceReferenceTypeVisitor(idFactory, isCrossResourceReference)
+
+			for _, def := range definitions {
+				// Skip Status types
+				// TODO: we need flags
+				if strings.Contains(def.Name().Name(), "_Status") {
+					result.Add(def)
+					continue
+				}
+
+				t, err := visitor.Visit(def.Type(), def.Name())
+				if err != nil {
+					return nil, errors.Wrapf(err, "visiting %q", def.Name())
+				}
+				result.Add(def.WithType(t))
+
+				// TODO: Remove types that have only a single field ID and pull things up a level? Will need to wait for George's
+				// TODO: Properties collapsing work for this.
+			}
+
+			return result, nil
+		})
+}
+
+type referencePair struct {
+	typeName astmodel.TypeName
+	propName astmodel.PropertyName
+}
+
+type crossResourceReferenceChecker func(typeName astmodel.TypeName, prop *astmodel.PropertyDefinition) bool
+
+type crossResourceReferenceTypeVisitor struct {
+	astmodel.TypeVisitor
+	// referenceChecker is a function describing what a cross resource reference looks like. It is overridable so that
+	// we can use a more simplistic criteria for tests.
+	isPropertyAnARMReference crossResourceReferenceChecker
+}
+
+func makeCrossResourceReferenceTypeVisitor(idFactory astmodel.IdentifierFactory, referenceChecker crossResourceReferenceChecker) crossResourceReferenceTypeVisitor {
+	visitor := crossResourceReferenceTypeVisitor{
+		isPropertyAnARMReference: referenceChecker,
+	}
+
+	transformResourceReferenceProperties := func(this *astmodel.TypeVisitor, it *astmodel.ObjectType, ctx interface{}) (astmodel.Type, error) {
+		typeName := ctx.(astmodel.TypeName)
+
+		var newProps []*astmodel.PropertyDefinition
+		for _, prop := range it.Properties() {
+			if visitor.isPropertyAnARMReference(typeName, prop) {
+				klog.V(4).Infof("Transforming \"%s.%s\" field into genruntime.ResourceReference", typeName, prop.PropertyName())
+				originalName := string(prop.PropertyName())
+				prop = makeResourceReferenceProperty(idFactory, prop)
+
+				// TODO: We could pass this information forward some other way?
+				// Add tag so that we remember what this field was before
+				prop = prop.WithTag(astmodel.ARMReferenceTag, originalName)
```

Comment:
Yeah, I simultaneously do not love how I am doing it here but also _kinda_ think it's neat, in that you can see on the actual property in the code that the reference maps to. It may get renamed as part of the whole dance so in some ways it's nice that it's there for reference in a similar way that the `json` annotates for "serializing to JSON" this annotates for "serializing to ARM"

Definitely open to others opinions on if they'd rather see this done differently though as I was torn when I wrote it too.